homeassistant/components/media_player/template_media_player.py

```diff
@@ -0,0 +1,198 @@
+    vol.Optional(CONF_INITIAL_VOLUME, default=50): cv.positive_int,
+    vol.Optional(CONF_VOLUME_STEP, default=5): cv.positive_int,
+    vol.Optional(CONF_VOLUME_STEP_DELAY, default=0.5): cv.positive_int,
+    vol.Optional(CONF_VOLUME_SET_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_VOLUME_MUTE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_VOLUME_UNMUTE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_SELECT_SOURCE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_SOURCE_NAMES, default=None): {cv.string: cv.string},
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the template media player platform."""
+    add_devices([
+        TemplateMediaPlayerDevice(
+            config.get(CONF_NAME),
+            config,
+            hass
+        )
+    ])
+
+
+class TemplateMediaPlayerDevice(MediaPlayerDevice):
+    """ Represents a generic device that delegates all operations to other services """
+
+    def __init__(self, name, config, hass):
+        """Initialize the device."""
+        self._name = name
+        self._actions = {
+            key: Script(hass, value) for key, value in config.items()
+            if key.endswith("_action")
+        }
+
+        self._sources = config.get(CONF_SOURCE_NAMES)
+        self._selected_source = None
+        self._state = STATE_OFF
+        self._volume = config.get(CONF_INITIAL_VOLUME)
+        self._volume_step = config.get(CONF_VOLUME_STEP)
+        self._volume_step_delay = config.get(CONF_VOLUME_STEP_DELAY)
+        self._is_muted = False
+
+        self._supported_features = 0
+        if CONF_ON_ACTION in self._actions:
+            self._supported_features |= SUPPORT_TURN_ON
+        if CONF_OFF_ACTION in self._actions:
+            self._supported_features |= SUPPORT_TURN_OFF
+        if (CONF_VOLUME_UP_ACTION in self._actions and
+            CONF_VOLUME_DOWN_ACTION in self._actions) or \
+            CONF_VOLUME_SET_ACTION in self._actions:
+            self._supported_features |= SUPPORT_VOLUME_SET
+        if CONF_VOLUME_MUTE_ACTION in self._actions and \
+           CONF_VOLUME_UNMUTE_ACTION in self._actions:
+            self._supported_features |= SUPPORT_VOLUME_MUTE
+        if self._sources and CONF_SELECT_SOURCE_ACTION in self._actions:
+            self._supported_features |= SUPPORT_SELECT_SOURCE
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        return self._supported_features
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return list(self._sources.values())
+
+    def turn_on(self):
+        """Turn on media player."""
+        self._state = STATE_ON
```

Comment:
We should allow users to configure a state template, and only fall back to optimistic updates if one isn't defined.